grr/server/grr_response_server/databases/spanner_utils.py

```diff
@@ -0,0 +1,486 @@
+  pass
+
+class Transaction(_Transaction):
+  """A wrapper around the PySpanner Transaction class."""
+  pass
+
+class Database:
+  """A wrapper around the PySpanner class.
+
+  The wrapper is supposed to streamline the usage of Spanner database through
+  an abstraction that is much harder to misuse. The wrapper will run retryable
+  queries through a transaction runner handling all brittle logic for the user.
+  """
+
+  _PYSPANNER_PARAM_REGEX = re.compile(r"@p\d+") 
+
+  def __init__(self, pyspanner: spanner_lib.Database, project_id: str) -> None:
+    super().__init__()
+    self._pyspanner = pyspanner
+    self.project_id = project_id
+  
+  def _parametrize(self, query: str, names: Iterable[str]) -> str:
+    match = self._PYSPANNER_PARAM_REGEX.search(query)
+    if match is not None:
+      raise ValueError(f"Query contains illegal sequence: {match.group(0)}")
+
+    kwargs = {}
+    for name in names:
+      kwargs[name] = f"@{name}"
+
+    return query.format(**kwargs)
+
+  def _get_param_type(self, value):
+    """
+    Infers the Google Cloud Spanner type from a Python value.
+
+    Args:
+        value: The Python value whose Spanner type is to be inferred.
+
+    Returns:
+        A google.cloud.spanner_v1.types.Type object, or None if the type
+        cannot be reliably inferred (e.g., for a standalone None value or
+        an empty list).
+    """
+    if value is None:
+        # Cannot determine a specific Spanner type from a None value alone.
+        # This indicates that the type is ambiguous without further schema context.
+        return None
+
+    py_type = type(value)
+
+    if py_type is int:
+        return param_types.INT64
+    elif py_type is float:
+        return param_types.FLOAT64
+    elif py_type is str:
+        return param_types.STRING
+    elif py_type is bool:
+        return param_types.BOOL
+    elif py_type is bytes:
+        return param_types.BYTES
+    elif py_type is datetime.date:
+        return param_types.DATE
+    elif py_type is datetime.datetime:
+        # Note: Spanner TIMESTAMPs are stored in UTC. Ensure datetime objects
+        # are timezone-aware (UTC) when writing data. This function only maps the type.
+        return param_types.TIMESTAMP
+    elif py_type is decimal.Decimal:
+        return param_types.NUMERIC
+    elif py_type is list:
+        if len(value) > 0:
+          return param_types.Array(self._get_param_type(value[0]))
+        else:
+          raise TypeError(f"Empty value for Python type: {py_type.__name__} for Spanner type conversion.")
+    else:
+        # Potentially raise an error for unsupported types or return None
+        # For a generic solution, raising an error for unknown types is often safer.
+        raise TypeError(f"Unsupported Python type: {py_type.__name__} for Spanner type conversion.")
+
+  def Transact(
+      self,
+      func: Callable[["Transaction"], _T],
+      txn_tag: Optional[str] = None,
+  ) -> List[Any]:
```

Comment:
```suggestion
  ) -> _T:
```

We should be returning the type of the function